confmap/resolver.go

```diff
@@ -195,6 +195,31 @@ func (mr *Resolver) Resolve(ctx context.Context) (*Conf, error) {
 	return retMap, nil
 }
 
+func escapeDollarSign(val any) any {
+	switch v := val.(type) {
+	case string:
+		return strings.ReplaceAll(v, "$$", "$")
+	case expandedValue:
+		v.Original = strings.ReplaceAll(v.Original, "$$", "$")
```

Comment:
@mx-psi do we need this escaping here? Isn't the Original field supposed to keep the string as is?

Reply:
@dmitryax Original is meant to keep the string representation. I think there are situations where an user would expect expansion to happen and situations where they wouldn't. 

#### Situation where an user expects expansion (and therefore escaping) to happen:

This is a bit of a contrived example but consider the following env vars and files:
- `URL_WITHOUT_PROTOCOL` set to `//example.com/api/v1/traces`
- `HTTP_URL` set to `http:${env:URL_WITHOUT_PROTOCOL}`
- a config file that sets the `otlp::endpoint` to `HTTP_URL`

In this situation, I would expect the OTLP endpoint to get `http://example.com/api/v1/traces`. 

Before #10618, this would fail with the `env` provider, since the contents of `HTTP_URL` would be parsed as a `map[string]any` with a single `http` key.

#### Situation where they do not expect expansion (or escaping) to happen

A password, that could have arbitrary contents

----

I don't know what to make out of this, we could have a third path so if a field is a `configopaque.String` we use the raw literal string (that's definitely doable with a bit of extra work and plumbing). I don't know if my imagination is good enough to cover all cases

Reply:
The tests still pass without the `OriginalValue` being escaped. Is there a scenario I could add to the e2e tests that would fail if `OriginalValue` isn't escape (or an inverse test if it should definitely not be escaped)